[win_chocolatey]Support for Side by side installs

alex3305 opened the pull request:
##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
win_chocolatey

##### ANSIBLE VERSION
```
ansible 2.2.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

##### SUMMARY
This commit adds support for side by side installations of the different
versions of the same Chocolatey package in Ansible.

Internally the -m flag will be set when running the choco.exe install *
command. This will make Chocolatey disregard the currently installed
version and thus not upgrade, but install it side-by-side.

This change will probably be unsupported by some packages and maybe there
will be a need to provide a path to Ansible's win_chocolatey module, so it
is easier to manage multiple side-by-side installations of the same
package.

lucalandi commented:
Can we get an update on this ? It's a really useful feature and I don't find any other way to inject it from win_chocolatey.

dagwieers commented:
I don't object to this functionality, but we first have to rebase it, and then fix it to follow the new standards for Windows modules.

Also, I don't see any option being passed to chocolatey to prevent the package to be upgraded. I am not sure this is going to work (but also did not test it).

dagwieers commented on lib/ansible/modules/windows/win_chocolatey.ps1:
```diff
@@ -40,6 +40,7 @@ $packageparams = Get-Attr -obj $params -name params -default $null
 $allowemptychecksums = Get-Attr -obj $params -name allow_empty_checksums -default "false" | ConvertTo-Bool
 $ignorechecksums = Get-Attr -obj $params -name ignore_checksums -default "false" | ConvertTo-Bool
 $ignoredependencies = Get-Attr -obj $params -name ignore_dependencies -default "false" | ConvertTo-Bool
+$allowmultipleversions = Get-Attr -obj $params -name allow_multiple_versions -default "false" | ConvertTo-Bool
```
Should be using the new `Get-AnsibleParam` functionality:
```powershell
$allowmultipleversions = Get-AnsibleParam -obj $params -name "allow_multiple_version" -type "bool" -default $false
```

alex3305 replied:
Done.

dagwieers commented on lib/ansible/modules/windows/win_chocolatey.py:
```diff
@@ -90,6 +90,12 @@
+    description:
+      - Allow multiple versions of the same package to be installed
+    require: false
+    default: false
```
Make this:
```yaml
    type: bool
    default: 'no'
```

alex3305 replied:
Done.

dagwieers commented on lib/ansible/modules/windows/win_chocolatey.py:
```diff
@@ -90,6 +90,12 @@
+  allow_multiple_versions:
+    description:
+      - Allow multiple versions of the same package to be installed
+    require: false
```
Remove this line. The default behaviour is that parameters are not required, unless you enable this.

alex3305 replied:
Done.

dagwieers commented on lib/ansible/modules/windows/win_chocolatey.py:
```diff
@@ -90,6 +90,12 @@
+      - Allow multiple versions of the same package to be installed
+    require: false
+    default: false
+    version_added: '2.3'
```
This should become:
```yaml
    version_added: '2.4'
```

alex3305 replied:
Done.

dagwieers commented:
@jborean93  Should we take over this PR ?

alex3305 commented:
@dagwieers I can still do this PR. Ansible slipped out of my mind during the holiday season.

dagwieers commented:
No problem, please go over the issues and rebase the code :-)

Thanks !

Commit:
Re-added side-by-side installation for win_chocolatey in 2.4

Since there were quite a lot of changes for the Ansible 2.3 and 2.4
releases in windows modules, this had to be rebased and resolved for
merging.

alex3305 commented:
@dagwieers Reworked all the requests.

dagwieers commented on lib/ansible/modules/windows/win_chocolatey.py:
```diff
@@ -94,6 +94,12 @@
     version_added: '2.1'
+  allow_multiple_versions:
+    description:
+      - Allow multiple versions of the same package to be installed
```
This needs a trailing dot. The style police is watching...

alex3305 replied:
Thank you, added.

dagwieers commented:
@jborean93 @jhawkesworth @nitzmahone This needs someone to test it.

I also wonder if the description of this functionality should come with a warning, or some information to the effect of this. (i.e. how do you remove a specific version afterwards ? how does this affect running a specific verison ? etc.)

alex3305 commented:
@dagwieers I've re-amended my commit a little to address your concerns regarding installing and uninstalling multiple versions. This is also highlighted in the chocolatey wiki as described:

* https://github.com/chocolatey/choco/wiki/CommandsInstall
* https://github.com/chocolatey/choco/wiki/CommandsUninstall

and should work out of the box with the `-m` flag.

dagwieers commented:
I still see 121 files changed :-/

alex3305 commented:
@dagwieers fixed that. Never used `rebase` before, sorry about that.

mattclay commented:
CI failure in integration tests:

```
{
"changed": false, 
"failed": true, 
"module_stderr": "At line:180 char:37\r\n+         [bool]$allowmultipleversions\r\n+                                     ~\r\nMissing ')' in function parameter list.\r\n\r\nAt line:156 char:1\r\n+ {\r\n+ ~\r\nMissing closing '}' in statement block.\r\n\r\nAt line:184 char:5\r\n+     )\r\n+     ~\r\nUnexpected token ')' in expression or statement.\r\n\r\nAt line:260 char:1\r\n+ }\r\n+ ~\r\nUnexpected token '}' in expression or statement.\r\n\r\nAt line:293 char:35\r\n+     if (Choco-IsInstalled $package) -and (-not $allowmultipleversions))\r\n+                                   ~\r\nMissing statement block after if ( condition ).\r\n\r\nAt line:263 char:1\r\n+ {\r\n+ ~\r\nMissing closing '}' in statement block.\r\n\r\nAt line:293 char:71\r\n+     if (Choco-IsInstalled $package) -and (-not $allowmultipleversions))\r\n+                                                                       ~\r\nUnexpected token ')' in expression or statement.\r\n\r\nAt line:373 char:1\r\n+ }\r\n+ ~\r\nUnexpected token '}' in expression or statement.\r\n\r\n\r\n    + CategoryInfo          : ParserError: (:) [], ParentContainsErrorRecordEx \r\n   ception\r\n    + FullyQualifiedErrorId : MissingEndParenthesisInFunctionParameterList\r\n \r\n\r\n", 
"module_stdout": "", 
"msg": "MODULE FAILURE", 
"rc": 1
}
```

Commit:
Re-added side-by-side installation for win_chocolatey in 2.4 [2]

Rebase apparently went completely wrong with this one..

Sorry guys, normally we merge :).

alex3305 commented:
@mattclay Fixed that for ya. The rebase went completely wrong. I had trouble rebasing the thing. Normally at work we use merge instead of rebase. Kinda confusing for the first time :). Should work now.

jborean93 commented on lib/ansible/modules/windows/win_chocolatey.ps1:
```diff
@@ -176,7 +177,7 @@ Function Choco-Upgrade
         [Parameter(Mandatory=$false, Position=10)]
         [int]$timeout,
-        [Parameter(Mandatory=$false, Position=11)]
+        [Parameter(Mandatory=$false, Position=12)]
```
Why did this change, should there be more lines to add `allowmultipleversions` to the 11th position?

jborean93 reviewed (changes requested):
I believe there is a call to `Choco-Upgrade` that needs to have this new parameter passed in now. Once you fix up the minor things I'll do a test before merging.

Commit:
Amended wrongful Choco-Upgrade parameter list

This also went wrong during my rebase, this commit fixes that issue.

jborean93 commented:
@alex3305, just tried this out locally and it doesn't work. When trying to install a newer version of putty side by side it just returns without a change. I believe the check `Choco-IsInstalled` is returning true and then just exists the function.
